[BUG] Build failure on 32-bit architecture (at least i386)

### Required prerequisites

- [X] I have read the documentation <https://optree.readthedocs.io>.
- [X] I have searched the [Issue Tracker](https://github.com/metaopt/optree/issues) that this hasn't already been reported. (comment there if it has.)

### What version of OpTree are you using?

0.11.0

### System information

The library was built from ports via 

### Problem description

It fails to build on i386 due to same size of different types.
It causes error of redefinition of functions.
Therefore, the build fails.

The full build log is [here](https://pkg-status.freebsd.org/beefy13/data/132i386-quarterly/40d461549d9f/logs/py39-optree-0.11.0.log).

### Reproducible example code

The build failure is as follows:
```
-- Build files have been written to: /wrkdirs/usr/ports/devel/py-optree/work-py39/optree-0.11.0/build/temp.freebsd-13.2-RELEASE-p11-i386-cpython-39
/usr/local/bin/cmake --build . --config Release --target _C --
[ 12%] Building CXX object src/CMakeFiles/_C.dir/optree.cpp.o
In file included from /wrkdirs/usr/ports/devel/py-optree/work-py39/optree-0.11.0/src/optree.cpp:25:
In file included from /wrkdirs/usr/ports/devel/py-optree/work-py39/optree-0.11.0/include/registry.h:29:
/wrkdirs/usr/ports/devel/py-optree/work-py39/optree-0.11.0/include/utils.h:229:19: error: redefinition of 'GET_ITEM_HANDLE<pybind11::tuple, int>'
inline py::handle GET_ITEM_HANDLE<py::tuple>(const py::handle& container, const int& item) {
                  ^
/wrkdirs/usr/ports/devel/py-optree/work-py39/optree-0.11.0/include/utils.h:221:19: note: previous definition is here
inline py::handle GET_ITEM_HANDLE<py::tuple>(const py::handle& container, const py::ssize_t& item) {
                  ^
/wrkdirs/usr/ports/devel/py-optree/work-py39/optree-0.11.0/include/utils.h:241:19: error: redefinition of 'GET_ITEM_HANDLE<pybind11::list, int>'
inline py::handle GET_ITEM_HANDLE<py::list>(const py::handle& container, const int& item) {
                  ^
/wrkdirs/usr/ports/devel/py-optree/work-py39/optree-0.11.0/include/utils.h:233:19: note: previous definition is here
inline py::handle GET_ITEM_HANDLE<py::list>(const py::handle& container, const py::ssize_t& item) {
                  ^
/wrkdirs/usr/ports/devel/py-optree/work-py39/optree-0.11.0/include/utils.h:259:19: error: redefinition of 'GET_ITEM_BORROW<pybind11::tuple, int>'
inline py::object GET_ITEM_BORROW<py::tuple>(const py::handle& container, const int& item) {
                  ^
/wrkdirs/usr/ports/devel/py-optree/work-py39/optree-0.11.0/include/utils.h:250:19: note: previous definition is here
inline py::object GET_ITEM_BORROW<py::tuple>(const py::handle& container, const py::ssize_t& item) {
                  ^
/wrkdirs/usr/ports/devel/py-optree/work-py39/optree-0.11.0/include/utils.h:273:19: error: redefinition of 'GET_ITEM_BORROW<pybind11::list, int>'
inline py::object GET_ITEM_BORROW<py::list>(const py::handle& container, const int& item) {
                  ^
/wrkdirs/usr/ports/devel/py-optree/work-py39/optree-0.11.0/include/utils.h:264:19: note: previous definition is here
inline py::object GET_ITEM_BORROW<py::list>(const py::handle& container, const py::ssize_t& item) {
                  ^
/wrkdirs/usr/ports/devel/py-optree/work-py39/optree-0.11.0/include/utils.h:292:19: error: redefinition of 'GET_ITEM_STEAL<pybind11::tuple, int>'
inline py::object GET_ITEM_STEAL<py::tuple>(const py::handle& container, const int& item) {
                  ^
/wrkdirs/usr/ports/devel/py-optree/work-py39/optree-0.11.0/include/utils.h:283:19: note: previous definition is here
inline py::object GET_ITEM_STEAL<py::tuple>(const py::handle& container, const py::ssize_t& item) {
                  ^
/wrkdirs/usr/ports/devel/py-optree/work-py39/optree-0.11.0/include/utils.h:306:19: error: redefinition of 'GET_ITEM_STEAL<pybind11::list, int>'
inline py::object GET_ITEM_STEAL<py::list>(const py::handle& container, const int& item) {
                  ^
/wrkdirs/usr/ports/devel/py-optree/work-py39/optree-0.11.0/include/utils.h:297:19: note: previous definition is here
inline py::object GET_ITEM_STEAL<py::list>(const py::handle& container, const py::ssize_t& item) {
                  ^
/wrkdirs/usr/ports/devel/py-optree/work-py39/optree-0.11.0/include/utils.h:328:13: error: redefinition of 'SET_ITEM<pybind11::tuple, int>'
inline void SET_ITEM<py::tuple>(const py::handle& container,
            ^
/wrkdirs/usr/ports/devel/py-optree/work-py39/optree-0.11.0/include/utils.h:316:13: note: previous definition is here
inline void SET_ITEM<py::tuple>(const py::handle& container,
            ^
/wrkdirs/usr/ports/devel/py-optree/work-py39/optree-0.11.0/include/utils.h:346:13: error: redefinition of 'SET_ITEM<pybind11::list, int>'
inline void SET_ITEM<py::list>(const py::handle& container,
            ^
/wrkdirs/usr/ports/devel/py-optree/work-py39/optree-0.11.0/include/utils.h:334:13: note: previous definition is here
inline void SET_ITEM<py::list>(const py::handle& container,
            ^
/wrkdirs/usr/ports/devel/py-optree/work-py39/optree-0.11.0/include/utils.h:578:9: error: no matching function for call to 'SET_ITEM'
        SET_ITEM<py::tuple>(fields, i, py::str(members[i].name));
        ^~~~~~~~~~~~~~~~~~~
/wrkdirs/usr/ports/devel/py-optree/work-py39/optree-0.11.0/include/utils.h:312:13: note: candidate template ignored: substitution failure [with Container = pybind11::tuple, Item = int]
inline void SET_ITEM(const py::handle& container, const Item& item, const py::handle& value) {
            ^
/wrkdirs/usr/ports/devel/py-optree/work-py39/optree-0.11.0/include/utils.h:672:26: error: no matching function for call to 'GET_ITEM_BORROW'
        py::object key = GET_ITEM_BORROW<py::list>(keys, i);
                         ^~~~~~~~~~~~~~~~~~~~~~~~~
/wrkdirs/usr/ports/devel/py-optree/work-py39/optree-0.11.0/include/utils.h:246:19: note: candidate template ignored: substitution failure [with Container = pybind11::list, Item = int]
inline py::object GET_ITEM_BORROW(const py::handle& container, const Item& item) {
                  ^
10 errors generated.
*** Error code 1
```

### Traceback

_No response_

### Expected behavior

_No response_

### Additional context

_No response_

--- a/.github/workflows/build.yml
+++ b/.github/workflows/build.yml
@@ -138,8 +138,8 @@ jobs:
         uses: pypa/cibuildwheel@v2.18
         env:
           CIBW_BUILD: ${{ env.CIBW_BUILD }}
-          CIBW_ARCHS_LINUX: auto64 aarch64 ppc64le s390x
-          CIBW_ARCHS_WINDOWS: auto64 ARM64
+          CIBW_ARCHS_LINUX: auto aarch64 ppc64le s390x
+          CIBW_ARCHS_WINDOWS: auto ARM64
           CIBW_ARCHS_MACOS: x86_64 arm64 universal2
         with:
           package-dir: .

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -13,7 +13,8 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ### Added
 
-- Add Linux ppc64le and s390x wheels by [@XuehaiPan](https://github.com/XuehaiPan) in [#25](https://github.com/metaopt/optree/pull/25).
+- Add 32-bit wheels for Linux and Windows by [@XuehaiPan](https://github.com/XuehaiPan) in [#141](https://github.com/metaopt/optree/pull/141).
+- Add Linux ppc64le and s390x wheels by [@XuehaiPan](https://github.com/XuehaiPan) in [#138](https://github.com/metaopt/optree/pull/138).
 - Add accessor APIs `tree_flatten_with_accessor` and `PyTreeSpec.accessors` by [@XuehaiPan](https://github.com/XuehaiPan) in [#108](https://github.com/metaopt/optree/pull/108).
 - Add submodule `optree.functools` by [@XuehaiPan](https://github.com/XuehaiPan) in [#134](https://github.com/metaopt/optree/pull/134).
 

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -51,7 +51,7 @@ if(MSVC)
         APPEND CMAKE_CXX_FLAGS
         " /Zc:preprocessor"
         " /experimental:external /external:anglebrackets /external:W0"
-        " /Wall /Wv:19.37"  # Visual Studio 2022 version 17.7
+        " /Wall /Wv:19.40"  # Visual Studio 2022 version 17.10
         # Suppress following warnings
         " /wd4365"  # conversion from 'type_1' to 'type_2', signed/unsigned mismatch
         " /wd4514"  # unreferenced inline function has been removed

--- a/include/treespec.h
+++ b/include/treespec.h
@@ -19,6 +19,7 @@ limitations under the License.
 
 #include <pybind11/pybind11.h>
 
+#include <algorithm>      // std::min
 #include <memory>         // std::unique_ptr
 #include <optional>       // std::optional, std::nullopt
 #include <string>         // std::string
@@ -38,7 +39,10 @@ using size_t = py::size_t;
 using ssize_t = py::ssize_t;
 
 // The maximum depth of a pytree.
-constexpr ssize_t MAX_RECURSION_DEPTH = 1000;
+#ifndef Py_C_RECURSION_LIMIT
+#define Py_C_RECURSION_LIMIT 1000
+#endif
+constexpr ssize_t MAX_RECURSION_DEPTH = std::min(1000, Py_C_RECURSION_LIMIT);
 
 // Test whether the given object is a leaf node.
 bool IsLeaf(const py::object &object,

--- a/include/utils.h
+++ b/include/utils.h
@@ -208,6 +208,9 @@ template <>
 inline ssize_t GET_SIZE<py::list>(const py::handle& sized) {
     return PyList_GET_SIZE(sized.ptr());
 }
+#ifndef PyDict_GET_SIZE
+#define PyDict_GET_SIZE PyDict_GetSize
+#endif
 template <>
 inline ssize_t GET_SIZE<py::dict>(const py::handle& sized) {
     return PyDict_GET_SIZE(sized.ptr());
@@ -222,90 +225,34 @@ inline py::handle GET_ITEM_HANDLE<py::tuple>(const py::handle& container, const 
     return PyTuple_GET_ITEM(container.ptr(), item);
 }
 template <>
-inline py::handle GET_ITEM_HANDLE<py::tuple>(const py::handle& container, const py::size_t& item) {
-    return PyTuple_GET_ITEM(container.ptr(), py::ssize_t_cast(item));
-}
-template <>
-inline py::handle GET_ITEM_HANDLE<py::tuple>(const py::handle& container, const int& item) {
-    return PyTuple_GET_ITEM(container.ptr(), py::ssize_t_cast(item));
-}
-template <>
 inline py::handle GET_ITEM_HANDLE<py::list>(const py::handle& container, const py::ssize_t& item) {
     return PyList_GET_ITEM(container.ptr(), item);
-}
-template <>
-inline py::handle GET_ITEM_HANDLE<py::list>(const py::handle& container, const py::size_t& item) {
-    return PyList_GET_ITEM(container.ptr(), py::ssize_t_cast(item));
-}
-template <>
-inline py::handle GET_ITEM_HANDLE<py::list>(const py::handle& container, const int& item) {
-    return PyList_GET_ITEM(container.ptr(), py::ssize_t_cast(item));
 }
 
 template <typename Container, typename Item>
 inline py::object GET_ITEM_BORROW(const py::handle& container, const Item& item) {
-    return container[item];
+    return py::reinterpret_borrow<py::object>(container[item]);
 }
 template <>
 inline py::object GET_ITEM_BORROW<py::tuple>(const py::handle& container, const py::ssize_t& item) {
     return py::reinterpret_borrow<py::object>(PyTuple_GET_ITEM(container.ptr(), item));
 }
 template <>
-inline py::object GET_ITEM_BORROW<py::tuple>(const py::handle& container, const py::size_t& item) {
-    return py::reinterpret_borrow<py::object>(
-        PyTuple_GET_ITEM(container.ptr(), py::ssize_t_cast(item)));
-}
-template <>
-inline py::object GET_ITEM_BORROW<py::tuple>(const py::handle& container, const int& item) {
-    return py::reinterpret_borrow<py::object>(
-        PyTuple_GET_ITEM(container.ptr(), py::ssize_t_cast(item)));
-}
-template <>
 inline py::object GET_ITEM_BORROW<py::list>(const py::handle& container, const py::ssize_t& item) {
     return py::reinterpret_borrow<py::object>(PyList_GET_ITEM(container.ptr(), item));
-}
-template <>
-inline py::object GET_ITEM_BORROW<py::list>(const py::handle& container, const py::size_t& item) {
-    return py::reinterpret_borrow<py::object>(
-        PyList_GET_ITEM(container.ptr(), py::ssize_t_cast(item)));
-}
-template <>
-inline py::object GET_ITEM_BORROW<py::list>(const py::handle& container, const int& item) {
-    return py::reinterpret_borrow<py::object>(
-        PyList_GET_ITEM(container.ptr(), py::ssize_t_cast(item)));
 }
 
 template <typename Container, typename Item>
 inline py::object GET_ITEM_STEAL(const py::handle& container, const Item& item) {
-    return container[item];
+    return py::reinterpret_steal<py::object>(container[item]);
 }
 template <>
 inline py::object GET_ITEM_STEAL<py::tuple>(const py::handle& container, const py::ssize_t& item) {
     return py::reinterpret_steal<py::object>(PyTuple_GET_ITEM(container.ptr(), item));
 }
 template <>
-inline py::object GET_ITEM_STEAL<py::tuple>(const py::handle& container, const py::size_t& item) {
-    return py::reinterpret_steal<py::object>(
-        PyTuple_GET_ITEM(container.ptr(), py::ssize_t_cast(item)));
-}
-template <>
-inline py::object GET_ITEM_STEAL<py::tuple>(const py::handle& container, const int& item) {
-    return py::reinterpret_steal<py::object>(
-        PyTuple_GET_ITEM(container.ptr(), py::ssize_t_cast(item)));
-}
-template <>
 inline py::object GET_ITEM_STEAL<py::list>(const py::handle& container, const py::ssize_t& item) {
     return py::reinterpret_steal<py::object>(PyList_GET_ITEM(container.ptr(), item));
-}
-template <>
-inline py::object GET_ITEM_STEAL<py::list>(const py::handle& container, const py::size_t& item) {
-    return py::reinterpret_steal<py::object>(
-        PyList_GET_ITEM(container.ptr(), py::ssize_t_cast(item)));
-}
-template <>
-inline py::object GET_ITEM_STEAL<py::list>(const py::handle& container, const int& item) {
-    return py::reinterpret_steal<py::object>(
-        PyList_GET_ITEM(container.ptr(), py::ssize_t_cast(item)));
 }
 
 template <typename Container, typename Item>
@@ -314,39 +261,15 @@ inline void SET_ITEM(const py::handle& container, const Item& item, const py::ha
 }
 template <>
 inline void SET_ITEM<py::tuple>(const py::handle& container,
-                                const ssize_t& item,
+                                const py::ssize_t& item,
                                 const py::handle& value) {
     PyTuple_SET_ITEM(container.ptr(), item, value.inc_ref().ptr());
 }
 template <>
-inline void SET_ITEM<py::tuple>(const py::handle& container,
-                                const py::size_t& item,
-                                const py::handle& value) {
-    PyTuple_SET_ITEM(container.ptr(), py::ssize_t_cast(item), value.inc_ref().ptr());
-}
-template <>
-inline void SET_ITEM<py::tuple>(const py::handle& container,
-                                const int& item,
-                                const py::handle& value) {
-    PyTuple_SET_ITEM(container.ptr(), py::ssize_t_cast(item), value.inc_ref().ptr());
-}
-template <>
 inline void SET_ITEM<py::list>(const py::handle& container,
-                               const ssize_t& item,
+                               const py::ssize_t& item,
                                const py::handle& value) {
     PyList_SET_ITEM(container.ptr(), item, value.inc_ref().ptr());
-}
-template <>
-inline void SET_ITEM<py::list>(const py::handle& container,
-                               const py::size_t& item,
-                               const py::handle& value) {
-    PyList_SET_ITEM(container.ptr(), py::ssize_t_cast(item), value.inc_ref().ptr());
-}
-template <>
-inline void SET_ITEM<py::list>(const py::handle& container,
-                               const int& item,
-                               const py::handle& value) {
-    PyList_SET_ITEM(container.ptr(), py::ssize_t_cast(item), value.inc_ref().ptr());
 }
 
 inline std::string PyRepr(const py::handle& object) {

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -109,7 +109,7 @@ optree = ['*.so', '*.pyd']
 # Wheel builder ################################################################
 # Reference: https://cibuildwheel.readthedocs.io
 [tool.cibuildwheel]
-archs = ["auto64"]
+archs = ["native"]
 skip = "*musllinux*"
 build-frontend = "build"
 build-verbosity = 3

--- a/setup.py
+++ b/setup.py
@@ -50,6 +50,8 @@ class cmake_build_ext(build_ext):  # noqa: N801
             archs = re.findall(r'-arch (\S+)', os.environ.get('ARCHFLAGS', ''))
             if archs:
                 cmake_args.append(f'-DCMAKE_OSX_ARCHITECTURES={";".join(archs)}')
+        elif platform.system() == 'Windows' and platform.architecture()[0] == '32bit':
+            cmake_args.append('-A=Win32')
 
         try:
             import pybind11

--- a/src/treespec/constructor.cpp
+++ b/src/treespec/constructor.cpp
@@ -78,7 +78,7 @@ template <bool NoneIsLeaf>
                                             std::vector<PyTreeSpec>& treespecs,
                                             std::string& register_namespace) {
         for (const py::object& child : children) {
-            if (!py::isinstance<PyTreeSpec>(child)) {
+            if (!py::isinstance<PyTreeSpec>(child)) [[unlikely]] {
                 std::ostringstream oss{};
                 oss << "Expected a(n) " << NodeKindToString(node) << " of PyTreeSpec(s), got "
                     << PyRepr(handle) << ".";
@@ -213,15 +213,15 @@ template <bool NoneIsLeaf>
                 throw std::runtime_error(oss.str());
             }
             node.arity = 0;
-            node.node_data = GET_ITEM_BORROW<py::tuple>(out, 1);
+            node.node_data = GET_ITEM_BORROW<py::tuple>(out, ssize_t(1));
             for (const py::handle& child :
-                 py::cast<py::iterable>(GET_ITEM_BORROW<py::tuple>(out, 0))) {
+                 py::cast<py::iterable>(GET_ITEM_BORROW<py::tuple>(out, ssize_t(0)))) {
                 ++node.arity;
                 children.emplace_back(py::reinterpret_borrow<py::object>(child));
             }
             verify_children(children, treespecs, registry_namespace);
             if (num_out == 3) [[likely]] {
-                py::object node_entries = GET_ITEM_BORROW<py::tuple>(out, 2);
+                py::object node_entries = GET_ITEM_BORROW<py::tuple>(out, ssize_t(2));
                 if (!node_entries.is_none()) [[likely]] {
                     node.node_entries = py::cast<py::tuple>(std::move(node_entries));
                     const ssize_t num_entries = GET_SIZE<py::tuple>(node.node_entries);

--- a/src/treespec/flatten.cpp
+++ b/src/treespec/flatten.cpp
@@ -144,14 +144,14 @@ bool PyTreeSpec::FlattenIntoImpl(const py::handle& handle,
                     throw std::runtime_error(oss.str());
                 }
                 node.arity = 0;
-                node.node_data = GET_ITEM_BORROW<py::tuple>(out, 1);
+                node.node_data = GET_ITEM_BORROW<py::tuple>(out, ssize_t(1));
                 for (const py::handle& child :
-                     py::cast<py::iterable>(GET_ITEM_BORROW<py::tuple>(out, 0))) {
+                     py::cast<py::iterable>(GET_ITEM_BORROW<py::tuple>(out, ssize_t(0)))) {
                     ++node.arity;
                     recurse(child);
                 }
                 if (num_out == 3) [[likely]] {
-                    py::object node_entries = GET_ITEM_BORROW<py::tuple>(out, 2);
+                    py::object node_entries = GET_ITEM_BORROW<py::tuple>(out, ssize_t(2));
                     if (!node_entries.is_none()) [[likely]] {
                         node.node_entries = py::cast<py::tuple>(std::move(node_entries));
                         const ssize_t num_entries = GET_SIZE<py::tuple>(node.node_entries);
@@ -338,16 +338,16 @@ bool PyTreeSpec::FlattenIntoWithPathImpl(const py::handle& handle,
                     throw std::runtime_error(oss.str());
                 }
                 node.arity = 0;
-                node.node_data = GET_ITEM_BORROW<py::tuple>(out, 1);
+                node.node_data = GET_ITEM_BORROW<py::tuple>(out, ssize_t(1));
                 py::object node_entries;
                 if (num_out == 3) [[likely]] {
-                    node_entries = GET_ITEM_BORROW<py::tuple>(out, 2);
+                    node_entries = GET_ITEM_BORROW<py::tuple>(out, ssize_t(2));
                 } else [[unlikely]] {
                     node_entries = py::none();
                 }
                 if (node_entries.is_none()) [[unlikely]] {
                     for (const py::handle& child :
-                         py::cast<py::iterable>(GET_ITEM_BORROW<py::tuple>(out, 0))) {
+                         py::cast<py::iterable>(GET_ITEM_BORROW<py::tuple>(out, ssize_t(0)))) {
                         recurse(child, py::int_(node.arity++));
                     }
                 } else [[likely]] {
@@ -355,7 +355,7 @@ bool PyTreeSpec::FlattenIntoWithPathImpl(const py::handle& handle,
                     node.arity = GET_SIZE<py::tuple>(node.node_entries);
                     ssize_t num_children = 0;
                     for (const py::handle& child :
-                         py::cast<py::iterable>(GET_ITEM_BORROW<py::tuple>(out, 0))) {
+                         py::cast<py::iterable>(GET_ITEM_BORROW<py::tuple>(out, ssize_t(0)))) {
                         if (num_children >= node.arity) [[unlikely]] {
                             throw std::runtime_error(
                                 "PyTree custom flatten function for type " +
@@ -500,9 +500,10 @@ py::list PyTreeSpec::FlattenUpTo(const py::object& full_tree) const {
             case PyTreeKind::DefaultDict: {
                 AssertExactStandardDict(object);
                 auto dict = py::reinterpret_borrow<py::dict>(object);
-                py::list expected_keys = (node.kind != PyTreeKind::DefaultDict
-                                              ? node.node_data
-                                              : GET_ITEM_BORROW<py::tuple>(node.node_data, 1));
+                py::list expected_keys =
+                    (node.kind != PyTreeKind::DefaultDict
+                         ? node.node_data
+                         : GET_ITEM_BORROW<py::tuple>(node.node_data, ssize_t(1)));
                 if (!DictKeysEqual(expected_keys, dict)) [[unlikely]] {
                     py::list keys = SortedDictKeys(dict);
                     auto [missing_keys, extra_keys] = DictKeysDifference(expected_keys, dict);
@@ -619,7 +620,7 @@ py::list PyTreeSpec::FlattenUpTo(const py::object& full_tree) const {
                         << " should return a 2- or 3-tuple, got " << num_out << ".";
                     throw std::runtime_error(oss.str());
                 }
-                py::object node_data = GET_ITEM_BORROW<py::tuple>(out, 1);
+                py::object node_data = GET_ITEM_BORROW<py::tuple>(out, ssize_t(1));
                 if (node.node_data.not_equal(node_data)) [[unlikely]] {
                     std::ostringstream oss{};
                     oss << "Mismatch custom node data; expected: " << PyRepr(node.node_data)
@@ -628,7 +629,7 @@ py::list PyTreeSpec::FlattenUpTo(const py::object& full_tree) const {
                 }
                 ssize_t arity = 0;
                 for (const py::handle& child :
-                     py::cast<py::iterable>(GET_ITEM_BORROW<py::tuple>(out, 0))) {
+                     py::cast<py::iterable>(GET_ITEM_BORROW<py::tuple>(out, ssize_t(0)))) {
                     ++arity;
                     agenda.emplace_back(py::reinterpret_borrow<py::object>(child));
                 }

--- a/src/treespec/traversal.cpp
+++ b/src/treespec/traversal.cpp
@@ -123,10 +123,10 @@ py::object PyTreeIter::NextImpl() {
                         << " should return a 2- or 3-tuple, got " << num_out << ".";
                     throw std::runtime_error(oss.str());
                 }
-                auto children = py::cast<py::tuple>(GET_ITEM_BORROW<py::tuple>(out, 0));
+                auto children = py::cast<py::tuple>(GET_ITEM_BORROW<py::tuple>(out, ssize_t(0)));
                 ssize_t arity = GET_SIZE<py::tuple>(children);
                 if (num_out == 3) [[likely]] {
-                    py::object node_entries = GET_ITEM_BORROW<py::tuple>(out, 2);
+                    py::object node_entries = GET_ITEM_BORROW<py::tuple>(out, ssize_t(2));
                     if (!node_entries.is_none()) [[likely]] {
                         const ssize_t num_entries =
                             GET_SIZE<py::tuple>(py::cast<py::tuple>(std::move(node_entries)));

--- a/src/treespec/treespec.cpp
+++ b/src/treespec/treespec.cpp
@@ -139,8 +139,8 @@ namespace optree {
 
         case PyTreeKind::DefaultDict: {
             py::dict dict{};
-            py::object default_factory = GET_ITEM_BORROW<py::tuple>(node.node_data, 0);
-            py::list keys = GET_ITEM_BORROW<py::tuple>(node.node_data, 1);
+            py::object default_factory = GET_ITEM_BORROW<py::tuple>(node.node_data, ssize_t(0));
+            py::list keys = GET_ITEM_BORROW<py::tuple>(node.node_data, ssize_t(1));
             if (node.original_keys) [[unlikely]] {
                 for (ssize_t i = 0; i < node.arity; ++i) {
                     dict[GET_ITEM_HANDLE<py::list>(node.original_keys, i)] = py::none();
@@ -307,11 +307,11 @@ namespace optree {
             auto expected_keys = py::reinterpret_borrow<py::list>(
                 root.kind != PyTreeKind::DefaultDict
                     ? root.node_data
-                    : GET_ITEM_BORROW<py::tuple>(root.node_data, 1));
+                    : GET_ITEM_BORROW<py::tuple>(root.node_data, ssize_t(1)));
             auto other_keys = py::reinterpret_borrow<py::list>(
                 other_root.kind != PyTreeKind::DefaultDict
                     ? other_root.node_data
-                    : GET_ITEM_BORROW<py::tuple>(other_root.node_data, 1));
+                    : GET_ITEM_BORROW<py::tuple>(other_root.node_data, ssize_t(1)));
             py::dict dict{};
             for (ssize_t i = 0; i < other_root.arity; ++i) {
                 dict[GET_ITEM_HANDLE<py::list>(other_keys, i)] = py::int_(i);
@@ -583,7 +583,7 @@ ssize_t PyTreeSpec::PathsImpl(Span& paths,
                 auto keys = py::reinterpret_borrow<py::list>(
                     root.kind != PyTreeKind::DefaultDict
                         ? root.node_data
-                        : GET_ITEM_BORROW<py::tuple>(root.node_data, 1));
+                        : GET_ITEM_BORROW<py::tuple>(root.node_data, ssize_t(1)));
                 for (ssize_t i = root.arity - 1; i >= 0; --i) {
                     cur -= recurse(cur, GET_ITEM_HANDLE<py::list>(keys, i));
                 }
@@ -686,7 +686,7 @@ ssize_t PyTreeSpec::AccessorsImpl(Span& accessors,
                 auto keys = py::reinterpret_borrow<py::list>(
                     root.kind != PyTreeKind::DefaultDict
                         ? root.node_data
-                        : GET_ITEM_BORROW<py::tuple>(root.node_data, 1));
+                        : GET_ITEM_BORROW<py::tuple>(root.node_data, ssize_t(1)));
                 for (ssize_t i = root.arity - 1; i >= 0; --i) {
                     cur -= recurse(cur, GET_ITEM_HANDLE<py::list>(keys, i), path_entry_type);
                 }
@@ -749,7 +749,8 @@ py::list PyTreeSpec::Entries() const {
             return py::getattr(root.node_data, Py_Get_ID(copy))();
         }
         case PyTreeKind::DefaultDict: {
-            return py::getattr(GET_ITEM_BORROW<py::tuple>(root.node_data, 1), Py_Get_ID(copy))();
+            return py::getattr(GET_ITEM_BORROW<py::tuple>(root.node_data, ssize_t(1)),
+                               Py_Get_ID(copy))();
         }
 
         default:
@@ -785,7 +786,8 @@ py::object PyTreeSpec::Entry(ssize_t index) const {
             return GET_ITEM_BORROW<py::list>(root.node_data, index);
         }
         case PyTreeKind::DefaultDict: {
-            return GET_ITEM_BORROW<py::list>(GET_ITEM_BORROW<py::tuple>(root.node_data, 1), index);
+            return GET_ITEM_BORROW<py::list>(GET_ITEM_BORROW<py::tuple>(root.node_data, ssize_t(1)),
+                                             index);
         }
 
         case PyTreeKind::None:
@@ -960,11 +962,11 @@ bool PyTreeSpec::IsPrefix(const PyTreeSpec& other, const bool& strict) const {
                 auto expected_keys = py::reinterpret_borrow<py::list>(
                     a->kind != PyTreeKind::DefaultDict
                         ? a->node_data
-                        : GET_ITEM_BORROW<py::tuple>(a->node_data, 1));
+                        : GET_ITEM_BORROW<py::tuple>(a->node_data, ssize_t(1)));
                 auto other_keys = py::reinterpret_borrow<py::list>(
                     b->kind != PyTreeKind::DefaultDict
                         ? b->node_data
-                        : GET_ITEM_BORROW<py::tuple>(b->node_data, 1));
+                        : GET_ITEM_BORROW<py::tuple>(b->node_data, ssize_t(1)));
                 py::dict dict{};
                 for (ssize_t i = 0; i < b->arity; ++i) {
                     dict[GET_ITEM_HANDLE<py::list>(other_keys, i)] = py::int_(i);
@@ -1171,9 +1173,9 @@ std::string PyTreeSpec::ToStringImpl() const {
             case PyTreeKind::DefaultDict: {
                 EXPECT_EQ(
                     GET_SIZE<py::tuple>(node.node_data), 2, "Number of auxiliary data mismatch.");
-                py::object default_factory = GET_ITEM_BORROW<py::tuple>(node.node_data, 0);
-                auto keys =
-                    py::reinterpret_borrow<py::list>(GET_ITEM_BORROW<py::tuple>(node.node_data, 1));
+                py::object default_factory = GET_ITEM_BORROW<py::tuple>(node.node_data, ssize_t(0));
+                auto keys = py::reinterpret_borrow<py::list>(
+                    GET_ITEM_BORROW<py::tuple>(node.node_data, ssize_t(1)));
                 EXPECT_EQ(GET_SIZE<py::list>(keys),
                           node.arity,
                           "Number of keys and entries does not match.");
@@ -1294,13 +1296,13 @@ std::string PyTreeSpec::ToString() const {
             if (node.kind == PyTreeKind::DefaultDict) [[unlikely]] {
                 EXPECT_EQ(
                     GET_SIZE<py::tuple>(node.node_data), 2, "Number of auxiliary data mismatch.");
-                py::object default_factory = GET_ITEM_BORROW<py::tuple>(node.node_data, 0);
+                py::object default_factory = GET_ITEM_BORROW<py::tuple>(node.node_data, ssize_t(0));
                 data_hash = py::hash(default_factory);
             }
             auto keys = py::reinterpret_borrow<py::list>(
                 node.kind != PyTreeKind::DefaultDict
                     ? node.node_data
-                    : GET_ITEM_BORROW<py::tuple>(node.node_data, 1));
+                    : GET_ITEM_BORROW<py::tuple>(node.node_data, ssize_t(1)));
             EXPECT_EQ(
                 GET_SIZE<py::list>(keys), node.arity, "Number of keys and entries does not match.");
             for (const py::handle& key : keys) {

--- a/src/treespec/treespec.cpp
+++ b/src/treespec/treespec.cpp
@@ -168,18 +168,33 @@ namespace optree {
 }
 
 /*static*/ py::object PyTreeSpec::GetPathEntryType(const Node& node) {
-    static const auto [SequenceEntry_ptr,
-                       MappingEntry_ptr,
-                       NamedTupleEntry_ptr,
-                       StructSequenceEntry_ptr] = []()
-        -> std::tuple<const py::object*, const py::object*, const py::object*, const py::object*> {
+    static const py::object* SequenceEntry_ptr = []() {
         const py::module_& mod = GetCxxModule();
         // NOTE: Use raw pointers to leak the memory intentionally to avoid py::object deallocation
         // and garbage collection.
-        return {new py::object{py::getattr(mod, "SequenceEntry")},
-                new py::object{py::getattr(mod, "MappingEntry")},
-                new py::object{py::getattr(mod, "NamedTupleEntry")},
-                new py::object{py::getattr(mod, "StructSequenceEntry")}};
+        // NOLINTNEXTLINE[cppcoreguidelines-owning-memory]
+        return new py::object{py::getattr(mod, "SequenceEntry")};
+    }();
+    static const py::object* MappingEntry_ptr = []() {
+        const py::module_& mod = GetCxxModule();
+        // NOTE: Use raw pointers to leak the memory intentionally to avoid py::object deallocation
+        // and garbage collection.
+        // NOLINTNEXTLINE[cppcoreguidelines-owning-memory]
+        return new py::object{py::getattr(mod, "MappingEntry")};
+    }();
+    static const py::object* NamedTupleEntry_ptr = []() {
+        const py::module_& mod = GetCxxModule();
+        // NOTE: Use raw pointers to leak the memory intentionally to avoid py::object deallocation
+        // and garbage collection.
+        // NOLINTNEXTLINE[cppcoreguidelines-owning-memory]
+        return new py::object{py::getattr(mod, "NamedTupleEntry")};
+    }();
+    static const py::object* StructSequenceEntry_ptr = []() {
+        const py::module_& mod = GetCxxModule();
+        // NOTE: Use raw pointers to leak the memory intentionally to avoid py::object deallocation
+        // and garbage collection.
+        // NOLINTNEXTLINE[cppcoreguidelines-owning-memory]
+        return new py::object{py::getattr(mod, "StructSequenceEntry")};
     }();
 
     switch (node.kind) {

